libsolutil/CommonData.h

```diff
@@ -232,6 +232,17 @@ std::set<K> keys(std::map<K, V> const& _map)
 	return applyMap(_map, [](auto const& _elem) { return _elem.first; }, std::set<K>{});
 }
 
+/// @returns a pointer to the entry of @a _map at @a _key, if there is one, and nullptr otherwise.
+template<typename MapType, typename KeyType>
+auto fetchValue(MapType&& _map, KeyType const& _key)
```

Comment:
How about letting the caller specify the default? Otherwise you'll soon find yourself writing code like
`if (getValueIfAny(m, 'key')) { ... }` :)

```suggestion
template<typename MapType, typename KeyType, typename ValueType>
auto valueOrDefault(MapType&& _map, KeyType const& _key, ValueType& _default)
```

Reply:
I can add both - we'll probably need both :-).

Reply:
That's harder than it looks to get right type-wise actually :-).

Reply:
Maybe using `MapType::value_type` (and assuming that `MapType` must be compatible with STL map that way) instead of having a separate `ValueType` would be good enough?

Reply:
Well, we will probably have to make some concessions somewhere. Ideally, I'd want all of this to work:
```
std::map<int, std::string> m{{1, "value"}};
std::string def = "defval";
std::string const cdef = "const string";
valueOrDefault(m, 2, def) = "newdefval"; // changes def to "newdefval", i.e. this takes ``std::string&`` as third argument
valueOrDefault(m, 1, def) = "newval"; // changes m[1] to "newval"
// valueOrDefault(m, 1, cdef) = "newval"; // compile time error: cannot assign to const ref
// valueOrDefault(m, 1) = "newval"; // compile time error: cannot assign to const ref
assert(valueOrDefault(m, 2) == std::string{}); // defaults the third argument to ``std::string const& val = {}`` and returns a const reference
assert(valueOrDefault(m, 1) == "newval");
assert(valueOrDefault(m, 2, "d") == "d"); // constructs the default value as ``std::string const& val = "key"`` and returns a const reference.
assert(valueOrDefault(m, 1, cdef) == "newval");
assert(valueOrDefault(m, 2, cdef) == "const string");
std::map<int, std::string> const cm{{1, "value"}};
// valueOrDefault(cm, 2, def) = "newdefval"; // compile time error: cannot assign to const ref
assert(valueOrDefault(cm, 1) == "value");
assert(valueOrDefault(cm, 2, "d") == "d");
assert(valueOrDefault(cm, 2) == std::string{});
```

Allowing all that correctly will definitely require quite some magic C++ template trickery. But I can see what I can do.
If we just take some random type, this will be annoying at any second use of it...

Reply:
Getting the value type itself is easy even generically - you could just use sth like
``decltype(_map.find(_key)->second)`` - the more interesting question is when to make sure this is a reference or a const reference and how to make sure the return type deduces correctly as well :-).

Reply:
`tryGet`?

Reply:
Another option would be to define our own `map` class that has a method `get`

Reply:
```
// valueOrDefault(m, 1, cdef) = "newval"; // compile time error: cannot assign to const ref
// valueOrDefault(m, 1) = "newval"; // compile time error: cannot assign to const ref
...
// valueOrDefault(cm, 2, def) = "newdefval"; // compile time error: cannot assign to const ref
```

You mean that you want these to work too or that you want them to result in errors?
I'd prefer the latter. I really do not expect functions like this to return anything mutable.

Reply:
The commented one *should* be errors. But the others should work.
I'd also want to avoid this copying stuff unintendedly. But I think I have a solution that allows all of that, bear with me ;-).